django/db/models/deletion.py

```diff
@@ -247,7 +281,31 @@ def collect(self, objs, source=None, nullable=False, collect_related=True,
                 if hasattr(field, 'bulk_related_objects'):
                     # It's something like generic foreign key.
                     sub_objs = field.bulk_related_objects(new_objs, self.using)
-                    self.collect(sub_objs, source=model, nullable=True)
+                    self.collect(sub_objs, source=model, nullable=True, fail_on_restricted=False)
+
+        if fail_on_restricted:
+            # During collection we may have found objects via RESTRICT
+            # relationships (restricted objects). To be able to delete them, we
+            # need to also have found them through a different CASCADE
+            # relationship. Fail here if any of them wasn't.
+            for model, instances in self.data.items():
+                self.clear_restricted_objects_from_set(model, instances)
+
+            for qs in self.fast_deletes:
+                self.clear_restricted_objects_from_queryset(qs.model, qs)
+
+            for model in self.restricted_objects:
+                for field in self.restricted_objects[model]:
+                    for obj in self.restricted_objects[model][field]:
+                        raise RestrictedError(
+                            "Cannot delete some instances of model '%s' because they are "
+                            "referenced through a restricted foreign key: '%s.%s'" % (
+                                model.__name__,
+                                obj.__class__.__name__,
+                                field.name,
+                            ),
+                            self.restricted_objects[model][field],
```

Comment:
Do we need to use `self.restricted_objects` multiple times? I would simplify this, e.g.
```python
for model, fields in self.restricted_objects.items():
    for field, objs in fields.items():
        for obj in objs:
            raise RestrictedError(
                "Cannot delete some instances of model '%s' "
                "because they are referenced through a restricted "
                "foreign key: '%s.%s'" % (
                    model.__name__,
                    obj.__class__.__name__,
                    field.name,
                ),
                objs,
            )
```